### PR TITLE
doc: add a doctest to demonstrate a separately-compiled regex set

### PR DESCRIPTION
**Pull request by cosmicexplorer**
It took me a while to understand the recommendation for how to get `Match` objects when using a `RegexSet`. I tried to rewrite that part and added a doctest to make it super clear!

### CODE REVIEW

**Review by BurntSushi — CHANGES_REQUESTED** [review]
Thanks! The substance looks good, but there are some details that I'd like to see fixed before merging.

Also, I do think these restrictions have a very good chance of being lifted in the future with new APIs on `RegexSet`. But it's still a long ways off so I agree it's worth clarifying the docs here for now.

File: src/re_set.rs [hunk 1]
```diff
@@ -62,10 +62,35 @@ $(#[$doc_regexset_example])*
...
+/// ```rust
+/// # use regex::{Regex, RegexSet};
+/// let patterns = ["foo", "bar"];
+/// // Both patterns will match different ranges of this string.
+/// let text = "barfoo";
+///
+/// // Compile a set matching any of our patterns.
+/// let set = RegexSet::new(&patterns).unwrap();
+/// // Compile each pattern independently.
+/// let regexen: Vec<_> = patterns.iter().map(|pat| Regex::new(pat).unwrap()).collect();
```
[the remark this answers is not in the document]
Comment (BurntSushi) [reply src/re_set.rs:78]: Errmm, yeah, not `set`. Whoops. `regexes` then. :)

**Comment by BurntSushi** [comment]
Also, please squash down to one commit.

File: src/re_set.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (cosmicexplorer) [reply src/re_set.rs]: Done! Thanks!

File: src/re_set.rs [hunk 2]
```diff
@@ -62,10 +62,35 @@ $(#[$doc_regexset_example])*
...
 ///
-/// Other features like finding the location of successive matches or their
-/// sub-captures aren't supported. If you need this functionality, the
-/// recommended approach is to compile each regex in the set independently and
-/// selectively match them based on which regexes in the set matched.
+/// Other features like finding match locations or capture groups
+/// aren't supported. If you need this functionality, the
+/// recommended approach is to compile each pattern in the set independently and
+/// scan the exact same input a second time with those independently compiled
+/// patterns:
```
[the remark this answers is not in the document]
Comment (cosmicexplorer) [reply src/re_set.rs]: Done, for both!

File: src/re_set.rs [hunk 3]
```diff
@@ -62,10 +62,35 @@ $(#[$doc_regexset_example])*
...
+///     // Dereference the match index to get the corresponding compiled pattern.
+///     .map(|match_idx| &regexen[match_idx])
+///     // To get match locations or any other info, we then have to search
+///     // the exact same text again, using our separately-compiled pattern.
+///     .map(|pat| pat.find(text).unwrap().as_str())
+///     .collect();
+///
+/// // Matches arrive in the order the constituent patterns were declared,
+/// // not the order they appear in the input.
+/// assert_eq!(["foo", "bar"].as_ref(), &matches);
```
[the remark this answers is not in the document]
Comment (cosmicexplorer) [reply src/re_set.rs]: I really appreciate this in-depth description! Thanks so much for clarifying!

File: src/re_set.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (cosmicexplorer) [reply src/re_set.rs]: `regexen` is terminology used in a few places in the emacs manual to collectively refer to multiple regex patterns.

File: src/re_set.rs [hunk 4]
```diff
@@ -62,10 +62,35 @@ $(#[$doc_regexset_example])*
...
+/// ```
```
[the remark this answers is not in the document]
Comment (cosmicexplorer) [reply src/re_set.rs:100]: Thank you! Super easy!

File: src/re_set.rs [hunk 5]
```diff
@@ -59,14 +59,13 @@ $(#[$doc_regexset_example])*
...
 ///
-/// Other features like finding match locations or capture groups
-/// aren't supported. If you need this functionality, the
-/// recommended approach is to compile each pattern in the set independently and
-/// scan the exact same input a second time with those independently compiled
-/// patterns:
+/// You cannot directly extract [`Match`][crate::Match] or [`Captures`][crate::Captures] objects
+/// from a regex set. If you need these operations, the recommended approach is to compile each
+/// pattern in the set independently and scan the exact same input a second time with those
+/// independently compiled patterns:
```
[the remark this answers is not in the document]
Comment (cosmicexplorer) [reply src/re_set.rs:70]: Awesome!

File: src/re_set.rs [hunk 6]
```diff
@@ -62,10 +62,35 @@ $(#[$doc_regexset_example])*
...
-/// sub-captures aren't supported. If you need this functionality, the
-/// recommended approach is to compile each regex in the set independently and
-/// selectively match them based on which regexes in the set matched.
+/// Other features like finding match locations or capture groups
+/// aren't supported. If you need this functionality, the
+/// recommended approach is to compile each pattern in the set independently and
+/// scan the exact same input a second time with those independently compiled
+/// patterns:
+/// ```rust
+/// # use regex::{Regex, RegexSet};
```
[the remark this answers is not in the document]
Comment (cosmicexplorer) [reply src/re_set.rs]: Ok, I have uncommented it! However, it looks like all the other examples in this *file* (`re_set.rs`) have this `use` statement commented out -- should I fix those, too? e.g. https://github.com/rust-lang/regex/blob/9aef5b1edc2a436244b936db53a03ed6d720e87e/src/re_set.rs#L425

**Comment by cosmicexplorer** [comment]
@BurntSushi:
> Also, I do think these restrictions have a very good chance of being lifted in the future with new APIs on RegexSet

That's awesome to hear!! Turned on notifs for this repo!

**Comment by cosmicexplorer** [comment]
Ping!

**Review by BurntSushi — APPROVED** [review]

**Pull request merged**